proposals/0000-sum-to-tag.md

```diff
@@ -0,0 +1,259 @@
+---
+author: David Feuer
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/369).
+
+# Add `sumToTag#` primop
+
+Add a `sumToTag#` primop indicating which sum alternative is
+represented by a particular unboxed sum value. For example,
+
+```haskell
+sumToTag# (# a | #) = 0#
+sumToTag# (# | | b #) = 2#
+```
+
+## Motivation
+
+It's sometimes useful to get the numerical value of an alternative.
+For example, I might write
+
+```haskell
+newtype Trit# = Trit# (# (##) | (##) | (##) #)
+
+tritToInt# :: Trit# -> Int#
+tritToInt# (Trit# (# (##)|| #)) = 0#
+tritToInt# (Trit# (# |(##)| #)) = 1#
+tritToInt# (Trit# (# ||(##) #)) = 2#
+```
```

Comment:
Do you have an example of where you might want to call `tritToInt#` in real code? Consistency with the API around boxed sums and `dataToTag#` is good enough for me to motivate this proposal, but I'm genuinely curious about where these functions become useful.

Reply:
No, I don't have anything immediate. But in #367 I speculate about possibly using unboxed sums to express, e.g., Boolean results from primops. If that should ever happen, then we may well want to be able to write things like

```haskell
3# +# sumToTag# (x ==# 12)
```

Even today, some people may choose to write functions producing such unboxed sums (to express the finite number of return values), and may not want them to have to inline those to get free conversion to `Int#` when they want to do arithmetic.